### PR TITLE
Allow imagebuilder use from a build in OpenShift

### PR DESCRIPTION
**Pull request by smarterclayton**
Add `imageOptimizationPolicy` on Docker build strategy.

WIP, still need to cgroup constrain image builder and start testing

[test]

### CODE REVIEW

**Comment by smarterclayton** [comment]
@bparees see for API changes

File: pkg/build/api/types.go [hunk 1]
```diff
@@ -570,6 +570,25 @@ type CustomBuildStrategy struct {
...
+
+const (
+	// ImageOptimizationNone will attempt to generate canonical Docker image.
+	ImageOptimizationNone ImageOptimizationPolicy = "None"
+
+	// ImageOptimizationSkipLayerCache is an experimental policy and will avoid creating
+	// unique layers for each dockerfile line, resulting in smaller images and saving time
+	// during creation. Some Dockerfile syntax is not fully supported - content added to
+	// a VOLUME by an earlier layer may have incorrect uid, gid, and filesystem permissions.
+	// If an unsupported setting is detected, the build will fail.
```
Comment (bparees) [inline pkg/build/api/types.go]: this seems to promise more than we can deliver in terms of compatibility.

i feel like it needs some additional wording about the fact that there may be other incompatibilities (especially if docker changes their behavior in the future) that we may not detect/warn about.

**Comment by bparees** [comment]
> @bparees see for API changes

api change seems ok, would like to see a little more descriptiveness around the implications of using the optimization.

File: pkg/build/api/types.go [hunk 2]
```diff
@@ -570,6 +570,25 @@ type CustomBuildStrategy struct {
...
+	ImageOptimizationSkipLayerCache ImageOptimizationPolicy = "SkipLayerCache"
+
+	// ImageOptimizationSkipLayerCacheAndWarn is the same as SkipLayerCache, but will only
+	// warn to the build output instead of failing when unsupported syntax is detected.
```
Comment (bparees) [inline pkg/build/api/types.go]: should explicitly declare it is experimental?

File: pkg/build/api/types.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/build/api/types.go]: those would be bugs though - you don't think experimental is sufficient?

File: pkg/build/api/types.go (hunk 2 above)
Comment (smarterclayton) [inline pkg/build/api/types.go]: Hrm, yeah.

File: pkg/build/api/types.go (hunk 1 above)
Comment (bparees) [inline pkg/build/api/types.go]: maybe.  if you're confident the VOLUME thing is the only known incompatibility i guess it's ok, i'm just playing my chicken little role.

Comment (smarterclayton) [inline pkg/build/api/types.go]: this is actually the only known issue.  Everything else works to the limit of our testing.

**Comment by smarterclayton** [comment]
[testextended][extended:core(builds|image_ecosystem)]

**Comment by smarterclayton** [comment]
Now rebased on top of ben's changes, I had to bring a bunch of dependencies back.  I also ended up duplicating pkg/util/docker/dockerfile into pkg/build/util/dockerfile

**Comment by csrwng** [comment]
The latest extended test failure is due to the jenkins image taking > 10min to pull:
```
2017-03-14 16:44:00.231481993 -0400 EDT - event for jenkins-1-v1b8j: {kubelet 172.18.4.98} Pulling: pulling image "openshift/jenkins-2-centos7@sha256:979c70f84466645d9ea3bc61624cfad15e3035fdfd8083413118074ca1e3fba6"
2017-03-14 16:53:59.553092993 -0400 EDT - event for jenkins: {deploymentconfig-controller } ReplicationControllerScaled: Scaled replication controller "jenkins-1" from 1 to 0
2017-03-14 16:53:59.558840412 -0400 EDT - event for jenkins-1: {replication-controller } SuccessfulDelete: Deleted pod: jenkins-1-v1b8j
````

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
This is ready for more substantial review (all tests that should pass are
now passing).  Still need some form of ACL / disablement.

**Comment by smarterclayton** [comment]
Added a new synthetic ACL check for access to optimization, gave it to no one by default.  I think this is ready for more extensive testing.

File: pkg/build/controller/strategy/docker.go [hunk 3]
```diff
@@ -21,13 +21,21 @@ type DockerBuildStrategy struct {
...
 // TODO: Make the Pod definition configurable
 func (bs *DockerBuildStrategy) CreateBuildPod(build *buildapi.Build) (*kapi.Pod, error) {
+	strategy := build.Spec.Strategy.DockerStrategy
+
+	// DEBUG ONLY: remove me
+	if strategy.ImageOptimizationPolicy == nil {
+		policy := buildapi.ImageOptimizationSkipLayers
+		strategy.ImageOptimizationPolicy = &policy
+	}
+	// DEBUG ONLY: remove me
```
Comment (bparees) [inline pkg/build/controller/strategy/docker.go]: remove me.

File: pkg/generate/git/repository.go [hunk 4]
```diff
@@ -364,7 +364,7 @@ func command(name, dir string, env []string, args ...string) (stdout, stderr str
 func timedCommand(timeout time.Duration, name, dir string, env []string, args ...string) (stdout, stderr string, err error) {
 	var stdoutBuffer, stderrBuffer bytes.Buffer
 
-	glog.V(4).Infof("Executing %s %s %s", strings.Join(env, " "), name, strings.Join(args, " "))
+	glog.V(4).Infof("Executing %s %s", name, strings.Join(args, " "))
```
Comment (bparees) [inline pkg/generate/git/repository.go]: why remove this information?

File: pkg/authorization/api/synthetic.go [hunk 5]
```diff
@@ -3,6 +3,7 @@ package api
 // Synthetic authorization endpoints
 const (
 	DockerBuildResource          = "builds/docker"
+	OptimizedDockerBuildResource = "builds/optimizeddocker"
```
Comment (smarterclayton) [inline pkg/authorization/api/synthetic.go]: In the future we're just going to take this away.  Nothing about the current path is less secure than Docker, it's just a hedge so we can selective enable it.

File: pkg/generate/git/repository.go (hunk 4 above)
Comment (smarterclayton) [inline pkg/generate/git/repository.go]: It's potentially a security leak, and also makes the entire output unreadable.  Print the env once at the top if necessary.

Comment (bparees) [inline pkg/generate/git/repository.go]: anyone who can view the build logs can generally access any information that might appear here.

Comment (smarterclayton) [inline pkg/generate/git/repository.go]: In general logs should contain no more security info than necessary, and ideally none (just enough to identify that you passed the right secret).  Reprinting the env hundreds of times makes the logs not useful.

**Comment by smarterclayton** [comment]
Added tests, dropped commits, changed logging, still have carry of openshift/source-to-image#711 but that'll be picked up next time it's rebased.

**Comment by smarterclayton** [comment]
Yay, all green!

**Comment by bparees** [comment]
lgtm.

**Comment by smarterclayton** [comment]
[merge]

**Pull request merged**